Add support for domain and local parts

meskio opened the pull request:
From draft-09 of the WKD spech the client should include the domain and
local part of the email address in the request. Let's add support for
that.

If the domain is not provided in the url path we parse it from the
requested domain. So we add multi-domain support even if the advanced
mode of WKD is not being used.

Closes: #4

meskio committed:
Add support for domain and local parts

From draft-09 of the WKD spech the client should include the domain and
local part of the email address in the request. Let's add support for
that.

If the domain is not provided in the url path we parse it from the
requested domain. So we add multi-domain support even if the advanced
mode of WKD is not being used.

Closes: #4

meskio commented:
I open this pull-req to discuss how can we add support for the advanced WKD method described from draft-09 on.

This proposal is breaking backwards compatibility, in pkg.go.dev it looks like there is no so many consumers of this library and all of them are yours.

I thought it might make more sense to keep things simple and break compatibility. But we could keep backward compatibility adding an extra function to the `Handler`:
```go
type Handler struct {
	Discover func(hash string) ([]*openpgp.Entity, error)
	DiscoverAdv func(hash, domain, local string) ([]*openpgp.Entity, error)
}

func (h Hander) discover(hash, domain, local string) ([]*openpgp.Entity, error) {
	if h.DiscoverAdv != nil {
		return h.DiscoverAdv(hash, domain, local)
	}
	return h.Discover(hash)
}
```
Or just drop completely the `domain` and `local` and keep with the same API. I'm thinking about using this library but for now I don't think I care so much about the `domain` and `local`.

I haven't tested this code, I decided to open the conversation about it first and see what way we want to go. I'll test it with gnupg and other wkd clients once we have clarity about it.

emersion commented on server.go:
```diff
@@ -47,9 +47,21 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if strings.HasPrefix(path, "/hu/") {
 		hash := strings.TrimPrefix(path, "/hu/")
-		h.serveDiscovery(w, r, hash)
+		domain := strings.TrimPrefix(r.URL.Host, "openpgp.")
```
This is the simple method, right? Then I don't think we should trim here. Otherwise it gives incorrect results for addresses like `root@openpgp.org`.

meskio replied:
Mmm, that should work . I mean to discover the key of *root@openpgp.org* the WKD client will do a query to `https://openpgp.openpgp.org/hu/<hash>?l=root`, so the `r.URL.Host` will be "openpgp.openpgp.org". `TrimPrefix` will only remove the first ocurrence of the prefix, so `strings.TrimPrefix("openpgp.openpgp.org", "openpgp.")` [will return "openpgp.org"](https://go.dev/play/p/RyT7Z1FA6br). Or am I missing something here?

emersion replied:
With the "simple method" described in the RFC, the domain name is used as-is, ie. no "openpgp" subdomain is used to make the request.

meskio replied:
I see, you are right. But if I understand correctly the RFC I don't need to trim anything, the simple method will always be directed to the original domain without additions, only the advanced method will be directed to a subdomain. I should pay more attention while reading RFCs. Thanks.

I'll remove the TrimPrefix.

emersion replied:
No worries :)

emersion commented on client.go:
```diff
@@ -31,7 +31,16 @@ func Discover(addr string) ([]*openpgp.Entity, error) {
 
-	url := "https://" + domain + Base + "/hu/" + hashLocal(local)
+	urlStart := "https://" + domain + Base
+	urlEnd := "/hu/" + hashLocal(local) + "?l=" + local
```
`local` needs to be url-escaped here

meskio replied:
True, good catch. I'll fix it.

emersion commented:
Yeah I don't care too much about breaking the API

meskio commented:
I've done a bunch of fixes and I think I have addressed your comments.

I've set up a dummy server using this library and tested that it works with gnupg, sequoia and this library. So hopefully now is properly implementing the standard.

meskio committed:
Use io.Reader as return value of the Discover function

Don't depend on golang.org/x/crypto/openpgp wich is deprecated and is
missing support for ed25519 keys. Now the Discover function returns a
reader should read the binary export of an openpgp key.

Closes: #8

meskio commented:
I added a commit adding support to #8. If you prefer them to be in two different pullreqs I'm happy to split them.

meskio committed:
Add context to the Discover function

Passing the context from the request down allows us to set timeouts
(by http.TimeoutHandler) and propagate them to the DB or whatever we use
to store the keys.

meskio commented:
I added another commit adding support for context, as it will be pretty useful for me to have timeouts from the request propagated to the LDAP I will use to store the keys. Let me know what you think.

meskio commented:
@emersion just a reminder, this mr is still open. If you don't have time to maintain this library tell me and I'll fork it for my needs.

Thank you.
